contest_service/src/main.rs

```diff
@@ -1,207 +1,296 @@
+            .get_contest_metadata_collection()
             .find_one(None, None)
             .await
             .map_err(|x| Status::internal(format!("{}", x)))?
-            .ok_or_else(|| Status::not_found("Default contest not found"))?)
+            .ok_or_else(|| Status::not_found("Contest metadata not found"))?)
     }
 }
 
+fn check_password(password: &str, user_doc: &Document) -> bool {
+    user_doc.get_str("password").map_or(false, |hash| {
+        PasswordHash::new(hash).map_or(false, |parsed_hash| {
+            argon2::Argon2::default()
+                .verify_password(password.as_bytes(), &parsed_hash)
+                .is_ok()
+        })
+    })
+}
+
 #[tonic::async_trait]
 impl Contest for ContestService {
     async fn auth_user(
         &self,
-        _request: Request<AuthUserRequest>,
+        request: Request<AuthUserRequest>,
     ) -> Result<Response<AuthUserResponse>, Status> {
-        let auth_user = _request.into_inner();
-        let auth_username = auth_user.name.clone();
-        let auth_password = auth_user.passw.clone();
+        let user = request.into_inner();
+        let username = user.username.clone();
         Ok(self
-            .get_contests_collection()
+            .get_users_collection()
             .find_one(
                 doc! {
-                    "users": doc! {
-                        "username": auth_username,
-                        "password": auth_password
-                    }
+                    "_id": username,
                 },
                 None,
             )
             .await
-            .map_err(|x| Status::internal(format!("{}", x)))? // TODO fix with error conversion
+            .map_err(|err| Status::internal(format!("{}", err)))? // TODO fix with error conversion
+            .filter(|user_doc| check_password(&user.password, user_doc))
             .map_or(
                 Response::new(AuthUserResponse {
-                    error: Some("Incorrect username or password".to_string()),
-                    username: None,
-                    fullname: None,
-                    token: None,
+                    response: Some(auth_user_response::Response::Failure(
+                        auth_user_response::Failure {
+                            error: "Incorrect username or password".to_string(),
+                        },
+                    )),
                 }),
-                |_| {
+                |user_doc| {
                     Response::new(AuthUserResponse {
-                        username: Some(auth_user.name),
-                        fullname: None,
-                        error: None,
-                        token: None,
+                        response: Some(auth_user_response::Response::Success(
+                            auth_user_response::Success {
+                                username: user_doc.get("_id").unwrap().to_string(),
+                                fullname: user_doc.get("fullName").unwrap().to_string(),
```

Comment:
Consider using the `get_string()` method in `mongodb::Document` in order to retrieve a string and not a generic object.

Reply:
Look at it now